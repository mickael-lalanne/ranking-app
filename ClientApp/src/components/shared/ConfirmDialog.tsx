import React, { useState, useEffect } from 'react';
import { css } from '@emotion/css';
import Dialog from '@mui/material/Dialog';
import DialogActions from '@mui/material/DialogActions';
import DialogContent from '@mui/material/DialogContent';
import DialogContentText from '@mui/material/DialogContentText';
import DialogTitle from '@mui/material/DialogTitle';
import TextField from '@mui/material/TextField';
import { Button } from '@mui/material';

const ConfirmDialog = ({
    title,
    open,
    cancelHandler,
    confirmHandler,
    content,
    textValidation,
}: {
    title: string;
    content: string;
    open: boolean;
    cancelHandler: () => void;
    confirmHandler: () => void;
    textValidation?: string;
}) => {
    const [disableConfirmButton, setDisableConfirmButton] = useState<boolean>(!!textValidation);

    // Reset disable state when open and textValidation props changed
    useEffect(() => {
        setDisableConfirmButton(!!textValidation);
    }, [textValidation]);

    // Check if the user can click on the Confirm button
    const onTextValidationChange = (
        event: React.ChangeEvent<HTMLInputElement>
    ): void => {
        setDisableConfirmButton(event.target.value !== textValidation);
    };

    const ShowTextValidation = (): React.JSX.Element | undefined => {
        if (textValidation) {
            return (
                <TextField
                    autoFocus
                    margin="dense"
                    id="name"
                    placeholder={textValidation}
                    fullWidth
                    variant="standard"
                    onChange={onTextValidationChange}
                />
            );
        }
    };

    return (
        <Dialog open={open} onClose={cancelHandler} style={{ padding: '5px'}}>
            <div className={dialog_container_style}>
                <DialogTitle>{title}</DialogTitle>
                <DialogContent>
                    <DialogContentText>{content}</DialogContentText>
                    {ShowTextValidation()}
                </DialogContent>
                <DialogActions>
                    <Button onClick={cancelHandler}>Cancel</Button>
                    <Button
                        disabled={disableConfirmButton && !!textValidation}
                        onClick={confirmHandler}
                    >
                        Confirm
                    </Button>
                </DialogActions>
            </div>
        </Dialog>
    );
};

export default ConfirmDialog;

/**
 * CSS STYLES
 */
const dialog_container_style = css({
    padding: '7.5px'
});
