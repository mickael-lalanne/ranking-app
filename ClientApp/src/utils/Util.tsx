import React from "react";

// All items that are not yet saved in the database starts with the following identifier
const TMP_ID_IDENTIFIER = '99999';

/**
 * Generate a temporary for items that are not yet saved in database
 * @returns {string} a random temporary id
 */
export const generateRandomId = (): string => {
    const randomNumber: number = Math.floor(Math.random() * 1000000000);
    return TMP_ID_IDENTIFIER + randomNumber;
};

/**
 * Check if an id is temporary
 * @param {string} id the id to check
 * @returns {boolean} true if the id is temporary
 */
export const isTemporaryId = (id: string): boolean => {
    return id.startsWith(TMP_ID_IDENTIFIER);
}

export interface ResizedImage {
    source: string; // the data URL of the image
    name: string;
}

/**
 * Resize an image while keeping its ratio
 * This function is called in the "onChange" event of input file
 * So we use the FileReader api to get the base image dimensions
 * @param {File} imageFile the image file to resize
 * @returns {Promise<ResizedImage>} an object containing the resized image as data URL and its name
 */
export const resizeImage = (imageFile: File): Promise<ResizedImage> => {
    const MAX_SIZE: number = 150;
    const reader = new FileReader();

    return new Promise ((resolve) => {
        reader.onload = () => {
            const tmpImg: HTMLImageElement = new Image();
    
            tmpImg.onload = () => {
                const resizeRatio = Math.min(MAX_SIZE / tmpImg.width, MAX_SIZE / tmpImg.height);

                // Initialize the canvas
                const canvas = document.createElement("canvas");
                const ctx = canvas.getContext('2d');
    
                // Set width and height
                canvas.width = tmpImg.width * resizeRatio;
                canvas.height = tmpImg.height * resizeRatio;
    
                // Draw image and export to a data-uri
                ctx!.drawImage(tmpImg, 0, 0, canvas.width, canvas.height);
    
                const dataURI = canvas.toDataURL();
                resolve({ source: dataURI, name: imageFile.name });
            };

            tmpImg.src = URL.createObjectURL(imageFile);
        }
        reader.readAsDataURL(imageFile);
    });

};

/**
 * @param array an array containing a createdAt field
 * @returns {Partial<{ createdAt: string }>[]} the same array but sorted by creation date
 */
export const sortByCreationDate = (array: Partial<{ createdAt: string }>[]) => {
    // Use slice() to avoid reference errors
    return array.slice().sort((a, b) => 
        new Date(a.createdAt!).getTime() - new Date(b.createdAt!).getTime()
    );
};

/**
 * Called in the template and tierlist editors
 * Returns a value for the MUI Tooltip component
 * @param {string?} content the text that should be displayed in the tooltip 
 * @returns {'' | React.JSX.Element | undefined} a span including the text if defined, otherwise undefined
 */
export const getTooltipTitleForSaveButtons = (
    content?: string
): '' | React.JSX.Element | undefined => {
    return content && <span style={{ fontSize: '15px'}}>{content}</span>;
}
