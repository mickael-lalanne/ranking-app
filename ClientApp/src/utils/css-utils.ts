export const ELEMENT_SIZE: string = '75px';
