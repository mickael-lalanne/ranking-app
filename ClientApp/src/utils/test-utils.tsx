import React, { PropsWithChildren } from 'react';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { setupStore } from '../app/store';
import type { AppStore, RootState } from '../app/store';
import { ThemeOptions, ThemeProvider, createTheme } from '@mui/material/styles';
import '@testing-library/jest-dom';
import { DndProvider } from 'react-dnd';
import { TouchBackend } from 'react-dnd-touch-backend';

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
export interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
    preloadedState?: Partial<RootState>;
    store?: AppStore;
}

// Mock MUI theme
const RANKING_APP_THEME_MOCK: ThemeOptions = {
    defaultRankingTheme: {
        primary: '',
        primaryLight: '',
        light: '',
        dark: '',
        info: '',
    },
};
const MockTheme = ({ children }: any) => {
    const theme = createTheme(RANKING_APP_THEME_MOCK);
    return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

export function renderWithProviders(
    ui: React.ReactElement,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        store = setupStore(preloadedState),
        ...renderOptions
    }: ExtendedRenderOptions = {}
) {
    function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
        return (
            <MockTheme>
                <Provider store={store}>
                    <DndProvider
                        backend={TouchBackend}
                        options={{ enableMouseEvents: true }}
                    >
                        {children}
                    </DndProvider>
                </Provider>
            </MockTheme>
        );
    }

    // Return an object with the store and all of RTL's query functions
    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
