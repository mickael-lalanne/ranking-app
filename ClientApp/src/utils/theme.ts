import { Theme, createTheme, alpha, ThemeOptions } from "@mui/material";

interface RankingTheme {
    primary: string;
    primaryLight: string;
    light: string;
    dark: string;
    info: string;
};
const PRIMARY_COLOR = '#BB2525';

declare module '@mui/material/styles' {
    interface Theme {
        defaultRankingTheme: {
            primary: string;
            primaryLight: string;
            light: string;
            dark: string;
            info: string;
        };
    }
    // allow configuration using `createTheme`
    interface ThemeOptions {
        defaultRankingTheme?: {
            primary: string;
            primaryLight: string;
            light: string;
            dark: string;
            info: string;
        };
    }
}

export const THEMES_COLORS: ThemeOptions['defaultRankingTheme'] = {
    primary: PRIMARY_COLOR,
    primaryLight: '#FF6969',
    light: '#FFF5E0',
    dark: '#141E46',
    info: '#9ED9EB'
}

export const RANKING_APP_THEME: Theme = createTheme({
    defaultRankingTheme: THEMES_COLORS,
    palette: {
        primary: {
            main: alpha(PRIMARY_COLOR, 0.7),
            light: alpha(PRIMARY_COLOR, 0.5),
            dark: alpha(PRIMARY_COLOR, 0.9),
        }
    }
});
