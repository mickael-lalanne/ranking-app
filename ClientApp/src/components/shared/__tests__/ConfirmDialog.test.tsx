import * as React from 'react';
import { renderWithProviders } from '../../../utils/test-utils';
import { screen } from '@testing-library/react';
import ConfirmDialog, { ConfirmDialogTypes } from '../ConfirmDialog';
import userEvent from '@testing-library/user-event';

const DEFAULT_PROPS: ConfirmDialogTypes = {
    title: 'My title',
    content: 'My content',
    open: true,
    cancelHandler: jest.fn(),
    confirmHandler: jest.fn(),
};
const textValidation: string = 'My text validation';

const ConfirmDialogTest = (props: ConfirmDialogTypes) => {
    return (
        <ConfirmDialog
            title={props.title}
            content={props.content}
            confirmHandler={props.confirmHandler}
            cancelHandler={props.cancelHandler}
            open={props.open}
            textValidation={props.textValidation}
        />
    );
};

describe('Screenshot', () => {
    it('takes screenshot in open state', () => {
        const component = renderWithProviders(ConfirmDialogTest(DEFAULT_PROPS));

        expect(component.baseElement).toMatchSnapshot('Open dialog');
    });

    it('takes screenshot in close state', () => {
        const component = renderWithProviders(
            ConfirmDialogTest({ ...DEFAULT_PROPS, open: false })
        );

        expect(component.baseElement).toMatchSnapshot('Close dialog');
    });
});

describe('Common', () => {
    it('validates confirm and cancel handlers', async () => {
        renderWithProviders(ConfirmDialogTest(DEFAULT_PROPS));

        // Click on the cancel button
        await userEvent.click(screen.getByText('Cancel'));
        expect(DEFAULT_PROPS.cancelHandler).toHaveBeenCalledTimes(1);
        // Click on the confirm button
        await userEvent.click(screen.getByText('Confirm'));
        expect(DEFAULT_PROPS.confirmHandler).toHaveBeenCalledTimes(1);
    });

    it('tests the text validator', async () => {
        renderWithProviders(
            ConfirmDialogTest({ ...DEFAULT_PROPS, textValidation })
        );

        // Type a different text than expected
        await userEvent.type(
            screen.getByPlaceholderText(textValidation),
            'a random text'
        );
        // The confirm button should be disabled
        expect(screen.getByText('Confirm').closest('button')?.disabled).toBe(
            true
        );

        // Type the text expected
        await userEvent.clear(screen.getByPlaceholderText(textValidation));
        await userEvent.type(
            screen.getByPlaceholderText(textValidation),
            textValidation
        );
        // Now, the confirm button should be enable
        expect(screen.getByText('Confirm').closest('button')?.disabled).toBe(
            false
        );
    });
});
