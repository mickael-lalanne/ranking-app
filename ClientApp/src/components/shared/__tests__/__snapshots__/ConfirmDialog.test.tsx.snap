// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screenshot takes screenshot in close state: Close dialog 1`] = `
<body
  style=""
>
  <div />
</body>
`;

exports[`Screenshot takes screenshot in open state: Open dialog 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
    role="presentation"
    style="padding: 5px;"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiModal-backdrop css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
      role="presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-labelledby=":r0:"
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm css-1t1j96h-MuiPaper-root-MuiDialog-paper"
        role="dialog"
      >
        <div
          class="css-55b3qz"
          data-cy="confirm-dialog"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root"
            id=":r0:"
          >
            My title
          </h2>
          <div
            class="MuiDialogContent-root css-ypiqx9-MuiDialogContent-root"
          >
            <p
              class="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiDialogContentText-root css-qfso29-MuiTypography-root-MuiDialogContentText-root"
            >
              My content
            </p>
          </div>
          <div
            class="MuiDialogActions-root MuiDialogActions-spacing css-knqc4i-MuiDialogActions-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Cancel
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
              data-cy="confirm-dialog-button"
              tabindex="0"
              type="button"
            >
              Confirm
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
